pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,309 @@
+			if apierrors.IsNotFound(err) && q.clock.Since(cmd.TimeAdded) > time.Second*5 {
+				return NewUnrecoverableError(err)
+			}
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		// We emitted this event when disruption was blocked on launching/termination.
+		// This does not block other forms of deprovisioning, but we should still emit this.
+		q.recorder.Publish(disruptionevents.Launching(nodeClaim, fmt.Sprintf("%s/%s", cmd.Method,
+			lo.Ternary(len(cmd.ReplacementKeys) > 0, "replace", "delete"))))
+		initializedStatus := nodeClaim.StatusConditions().GetCondition(v1beta1.Initialized)
+		if !initializedStatus.IsTrue() {
+			q.recorder.Publish(disruptionevents.WaitingOnReadiness(nodeClaim))
+			waitErrs[i] = fmt.Errorf("node claim not initialized")
+			continue
+		}
+		cmd.ReplacementKeys[i].Initialized = true
+		// Subtract the last initialization time from the time the command was added to get initialization duration.
+		initLength := initializedStatus.LastTransitionTime.Inner.Time.Sub(nodeClaim.CreationTimestamp.Time).Seconds()
+		disruptionReplacementNodeClaimInitializedHistogram.Observe(initLength)
+	}
+	// If we have any errors, don't continue
+	if err := multierr.Combine(waitErrs...); err != nil {
+		return fmt.Errorf("waiting for replacement initialization, %w", err)
+	}
+
+	// All replacements have been provisioned.
+	// All we need to do now is get a successful delete call for each node claim,
+	// then the termination controller will handle the eventual deletion of the nodes.
+	var multiErr error
+	for i := range cmd.Candidates {
+		candidate := cmd.Candidates[i]
+		q.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, fmt.Sprintf("%s/%s", cmd.Method,
+			lo.Ternary(len(cmd.ReplacementKeys) > 0, "replace", "delete")))...)
+		if err := nodeclaim.Delete(ctx, q.kubeClient, candidate.NodeClaim); err != nil {
+			multiErr = multierr.Append(multiErr, client.IgnoreNotFound(err))
+		} else {
+			nodeclaim.TerminatedCounter(cmd.Candidates[i].NodeClaim, cmd.Method).Inc()
+		}
+	}
+	// If there were any deletion failures, we should requeue.
+	// In the case where we requeue, but the timeout for the command is reached, we'll mark this as a failure.
+	if multiErr != nil {
+		return fmt.Errorf("terminating nodeclaims, %w", multiErr)
+	}
+	return nil
+}
+
+// Add adds commands to the Queue
+// Each command added to the queue should already be validated and ready for execution.
+func (q *Queue) Add(cmd *Command) error {
+	providerIDs := lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+		return s.ProviderID()
+	})
+	// First check if we can add the command.
+	if err := q.CanAdd(providerIDs...); err != nil {
+		return fmt.Errorf("adding command, %w", err)
+	}
+	for _, candidate := range cmd.Candidates {
+		q.ProviderIDToCommand[candidate.ProviderID()] = cmd
+	}
+	// Idempotently mark for deletion
+	q.cluster.MarkForDeletion(providerIDs...)
```

Comment:
Should we be doing this in here or just assume that it's done ahead of time? I think in this case we're doing this twice which means that we will be holding the lock twice

Reply:
I was thinking idempotently this wouldn't make a difference, but holding the lock is a good reason not to do this twice.